src/include/catalog/postgres/pg_constraint.h

```diff
@@ -51,4 +51,172 @@ enum class ConstraintType : char {
+   * set fk related parameters to zero or false to occupy space
+   */
+  PG_Constraint(constraint_oid_t con_id, namespace_oid_t namespace_id, ConstraintType con_type,
+                table_oid_t con_table_id, col_oid_t col_id) {
+    oid_ = con_id;
+    namespace_id_ = namespace_id;
+    type_ = con_type;
+    table_id_ = con_table_id;
+    col_id_ = col_id;
+
+    // void FK attr
+    fk_ref_col_id_ = col_oid_t(0);
+    fk_ref_table_id_ = table_oid_t(0);
+    fk_update_cascade_ = false;
+    fk_delete_cascade_ = false;
+  }
+
+  /**
+   * Constructor for FK constraint
+   * requires all attribbutes including those for FK to be set
+   */
+  PG_Constraint(constraint_oid_t con_id, namespace_oid_t namespace_id, ConstraintType con_type,
+                table_oid_t con_table_id, col_oid_t col_id, table_oid_t ref_table, col_oid_t ref_col,
+                bool update_cascade, bool delete_cascade) {
+    oid_ = con_id;
+    namespace_id_ = namespace_id;
+    type_ = con_type;
+    table_id_ = con_table_id;
+    col_id_ = col_id;
+
+    // void FK attr
+    fk_ref_col_id_ = ref_col;
+    fk_ref_table_id_ = ref_table;
+    fk_update_cascade_ = update_cascade;
+    fk_delete_cascade_ = delete_cascade;
+  }
+};
+
+/**
+ * Manager class for pg_constraints
+ * responsible for creating and registering a constraint in the system
+ *  create this when database start 
+ */
+class PG_Constraint_Manager {
+ public:
+  // <oid_, constraint> map to get constraint from its oid
+  std::unordered_map<constraint_oid_t, PG_Constraint *> con_id_map_;
+  // the current highest oid available to be assigned to the new one
+  constraint_oid_t cur_oid_;
+  // queue used to record voided constraint id to be assigned to new constraints
+  std::queue<constraint_oid_t> unused_con_oid_;
+
+  PG_Constraint_Manager() {
+    cur_oid_ = constraint_oid_t(1);  // starting with 1 as 0 is the default voided id
+  }
+  // create a new normal constraint
+  PG_Constraint *GetNewConstraint(namespace_oid_t namespace_id, ConstraintType con_type, table_oid_t con_table_id,
+                                  col_oid_t col_id) {
+    constraint_oid_t tmp_oid;
+    PG_Constraint *con;
+    latch_.Lock();
+    if (!unused_con_oid_.empty()) {
+      tmp_oid = unused_con_oid_.front();
+      unused_con_oid_.pop();
+    } else {
+      tmp_oid = cur_oid_;
+      cur_oid_++;
+    }
+    con = new PG_Constraint(tmp_oid, namespace_id, con_type, con_table_id, col_id);
+    con_id_map_.emplace(tmp_oid, con);
+    latch_.Unlock();
+    return con;
+  }
+
+  // create a new FK constraint
+  PG_Constraint *GetNewFKConstraint(namespace_oid_t namespace_id, ConstraintType con_type, table_oid_t con_table_id,
+                                    col_oid_t col_id, table_oid_t ref_table, col_oid_t ref_col, bool update_cascade,
+                                    bool delete_cascade) {
+    constraint_oid_t tmp_oid;
+    PG_Constraint *con;
+    latch_.Lock();
+    if (!unused_con_oid_.empty()) {
```

Comment:
This part of code seems duplicated with the part in GetNewConstraint.
Could consider putting it in a helper function for finding next con_oid.